Add telescope positions to database and improve ArrayModel

GernotMaier opened the pull request:
**This PR includes major additions to the DB model and requires to set `SIMTOOLS_DB_SIMULATION_MODEL='Staging-CTA-Simulation-Model-v0-2-0'` in your `.env` file**

This PR simplifies and improves the ArrayModel module and uses now the newly added telescope positions from the DB.

Note that this PR is more complex than initially expected due to inter dependencies of layouts with the array configuration handling.

In detail:

- ArrayModel consists now of a SiteModel and a dict of TelescopeModels (this can be extended to add calibration devices)
- the TelescopeModel has been extended to include for all telescopes their position (in CORSIKA and UTM coordinates; formally we need only one, but for a start we keep both)
- the ArrayLayout class has been removed as dependency from all model and corsika/sim_telarray configuration classes (as all the necessary information for site and array is in ArrayModel). This means that ArrayLayout and TelescopePosition is used for coordinate transformation only and not anymore for the definition of the array model or writing of the configuration files. This simplifies significantly the corsika and simtel config modules (as all they need is the ArrayModel for a complete definition of the Observatory).
- moved "knowledge" about telescope models (e.g., if there is a design or explicit model defined) into the model class. This explains the change of parameter names from `telescope_model_name` to `telescope_name` (the assumption is that this is an internal item of the model classes which does not need to be known throughout the code)
- add application to write json files for each telescope positions from a table (ecsv file) of telescope positions (`simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py`)
- Simplified array and simulation configuration (e.g., tests/resources/array_config_test.yml or ./tests/resources/prod_multi_config_test_North.yml) and its usage. 
- add possibility to run corsika with simulate_prod (use `--simulator corsika`). The duplicated application `production.py` has been removed (this was necessary to include in this PR due to the changes in the array layout / array configuration 
- change of how telescope model is read from DB (see also issue #879): first read design model and then update the design model with the model of the specified telescopes. This allows with respect to this PR to define for all telescopes the positions, without having to copy the complete model for each telescope
- update the test layout files in test resources with the prod6 telescope positions
- note that telescopes positions for no in the DB yet for prod5 South (this is why those integration tests are skipped)

Side note:
- aware about the complain on missing tests for L277 in array_model. I just don't know how this can happen - either Raul or Orel put that in for a reason...

There are a couple of follow issues listed in https://github.com/gammasim/simtools/milestone/13

VictorBarbosaMartins commented on simtools/simtel/simtel_config_writer.py:
```diff
@@ -164,13 +162,15 @@ def write_array_config_file(self, config_file_path, layout, telescope_model, sit
 
             # Default telescope - 0th tel in telescope list
-            tel_config_file = telescope_model[0].get_config_file(no_export=True).name
+            # TODO check if this is correct - probably not!
```
Not sure whether this was fogoten here or meant for the future?

VictorBarbosaMartins commented on simtools/simtel/simtel_runner_array.py:
```diff
@@ -200,6 +200,7 @@ def has_file(self, file_type, run_number, mode="out"):
 
         info_for_file_name = self.get_info_for_file_name(run_number)
         run_sub_file = self.get_file_name(file_type, **info_for_file_name, mode=mode)
+        self._logger.debug(f"Checking if {run_sub_file} exists")
```
as in the other comment, it is not clear where the check is done. I would suggest either to change the message or the name of the function (getting file name and checking if file exists is not the same)

VictorBarbosaMartins commented on tests/resources/array_config_test.yml:
```diff
@@ -1,18 +1,20 @@
+  default:
+      LSTN: 'design'
+      MSTN: 'design'
+  LSTN-01:
+      name: "01"
+  LSTN-02:
+      name: "02"
+  LSTN-03:
+      name: "03"
+  LSTN-04:
+      name: "04"
```
I thought we would solve already the issue with the name, which is given twice here (as we spoke last week). Is there a plan to do that ? And why are the names different, i.e. "LSTN-01" is "01" and not "LSTN-01"?

GernotMaier replied:
As mentioned above: array config needs a rework (not as part of the PR).

VictorBarbosaMartins commented on tests/resources/telescope_positions-South-ground.ecsv:
```diff
@@ -54,6 +54,10 @@
 #     REFERENCE: {VERSION: 1.0.0}
 # schema: astropy-2.0
 telescope_name position_x position_y position_z geo_code
+LSTS-01     -20.63     -64.84      34.00  A1A
+LSTS-02      80.04      -0.77      29.00  B0A
+LSTS-03     -19.39      65.22      31.00  A3A
+LSTS-04    -105.06       1.15      33.00  Y0B
```
were telescopes missing here or did you expand to an upgraded array?

GernotMaier replied:
Beta contains LSTs. We had Alpha before.

VictorBarbosaMartins reviewed (changes requested):
Thank for this major update. I have mostly minor comments but quite a lot of them. I hope they help to improve the PR.

Point 1:
In general, I have the impression that the ArrayLayout class, as we spoke, is becoming slowly obsolete (and with that the tool make_regular_arrays). I suggest opening an issue on that to focus on finalizing the transition to the array model class (without losing functionality).

Point 2:
I suggest we already add the positions of the telescopes for prod5 to the DB as well. we have a stable environment at the moment running tests for prod5, but if we merge this branch without updating prod5 parameters, the tests will continue to be skipped. With this new db tool, it should be easy to do that, no?

GernotMaier commented on simtools/applications/db_development_tools/add_model_parameters_from_repository_to_db.py:
```diff
@@ -1,8 +1,8 @@
+    This application adds all parameters found in a repository to a new database.
 
-    Generates a new data with all required collections.
+    Generates a new database with all required collections.
```
I have looked into this and I think we don't need to update it: if the database exists, it will add/update the existing database. So this is fine, as we sometimes add new parameters or updates to parameters.

GernotMaier commented on simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py:
```diff
@@ -0,0 +1,188 @@
+
+    input : str
+        List of array element positions.
+
```
We are not using ruff yet...

But correct, I've removed the empty lines.

GernotMaier commented on simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py:
```diff
@@ -0,0 +1,188 @@
+
+    Command line arguments
+
+    input : str
```
Good point, improved the docstring.

GernotMaier commented on simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py:
```diff
@@ -0,0 +1,188 @@
+    input : str
+        List of array element positions.
+
+    output_path : str
+        Path of local copy of model parameter repository.
```
(puzzled a bit, but then realized that `now` should be `not`).

It should say `repository_path`, I've corrected it.

GernotMaier commented on simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py:
```diff
@@ -0,0 +1,188 @@
+    output_path : str
+        Path of local copy of model parameter repository.
+
+    model_version : str
+        Model version.
+
+    site : str
+        Observatory site.
+
+    coordinate_system : str
+        Coordinate system of array element positions (ground or utm).
```
model_version and site are tested in ArrayModel. I've added one check to make sure that only ground and utm is used for the coordinate system.

GernotMaier commented on simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py:
```diff
@@ -0,0 +1,188 @@
+
+    config = configurator.Configurator(label=label, description=description)
+    config.parser.add_argument(
+        "--input",
```
Done.

GernotMaier commented on simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py:
```diff
@@ -0,0 +1,188 @@
+        required=False,
+    )
+    config.parser.add_argument(
+        "--repository_path",
```
Fixed.

GernotMaier commented on simtools/applications/db_development_tools/write_array_elements_positions_to_repository.py:
```diff
@@ -0,0 +1,188 @@
+
+def write_utm_array_elements_to_repository(args_dict, logger):
+    """
+    Write UTM position of array elements to model repository.
+    Read array element positions from file.
```
Yes and no. The main intention of the function is to write, this is why I put it first (I think the convention is that a docstring starts with the main point of a function).

GernotMaier commented on simtools/applications/sim_showers_for_trigger_rates.py:
```diff
@@ -137,18 +137,22 @@ def main():
-        "azimuth": args_dict["azimuth"] * u.deg,
-        "viewcone": 10 * u.deg,
-        "cscat": [20, 1500 * u.m, 0],
+        "common": {
```
The whole array configuration is a bit chaotic and needs proper documents and a redo. This is planned to be done with issue #911

GernotMaier commented on simtools/corsika/corsika_config.py:
```diff
@@ -594,3 +571,27 @@ def _convert_to_quantities(self, value_args):
+            Piece of text to be added to the CORSIKA input file.
+        """
+
+        corsika_list = ""
```
Changed it, as you are right that it is confusing to mix python types with common language. I still use the word `list`, as it is really a list (just not a python list).

GernotMaier commented on simtools/corsika/corsika_runner.py:
```diff
@@ -396,6 +373,7 @@ def has_file(self, file_type, run_number=None, mode="out"):
 
         info_for_file_name = self.get_info_for_file_name(run_number)
         run_sub_file = self.get_file_name(file_type, **info_for_file_name, mode=mode)
+        self._logger.debug(f"Checking if {run_sub_file} exists")
```
`return Path(run_sub_file).is_file()` is the next line and checks that the file exists.

I've added this statement, as during debugging it is useful to see where the run files are written to (our directory structures can be complicated).

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -113,8 +114,11 @@ def get_model_parameters(
-        Get parameters from either MongoDB or simulation model repository for a specific
-        array element, i.e. telescopes, calibration_device.
+        Get parameters from MongoDB or simulation model repository for an array element.
+        An array element can be e.g., a telescope or a calibration device.
```
No, it is not (and there are other places to be changed). Tobias opened in issue on this to go through it systematically: #871

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -262,31 +265,16 @@ def _get_model_parameters_mongo_db(
-            )
-
-        return _pars
+        self._logger.debug(f"Getting {telescope_model_name} parameters from MongoDB")
+        return self.read_mongo_db(
+            db_name,
+            telescope_model_name,
+            model_version,
+            collection_name=collection,
+            run_location=None,
+            write_files=False,
+            only_applicable=only_applicable,
+        )
```
Agree - removed it.

(this has a history and there was more in this function in the past)

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1190,6 +1223,8 @@ def get_telescope_db_name(self, telescope_name, model_version, collection):
         model_version: str
             Which version to get the telescopes
+        collection: str
+            collection of array element (e.g. telescopes, calibration_devices)
```
Yes, this is issue #918. We need first get the available collections from the DB and then check their availability.

GernotMaier commented on simtools/db/db_handler.py:
```diff
@@ -1216,7 +1251,7 @@ def get_telescope_db_name(self, telescope_name, model_version, collection):
             return _design_name
 
-        self._logger.error("Telescope %s not found in the database.", telescope_name)
+        self._logger.debug("Telescope %s not found in the database.", telescope_name)
         raise ValueError
```
Removed it (this is one of the function which are usually called in try / except statements.)

GernotMaier commented on simtools/model/array_model.py:
```diff
@@ -18,190 +16,189 @@ class InvalidArrayConfigData(Exception):
 
-    @property
-    def number_of_telescopes(self):
+    def _initialize(self, site, array_elements_file, parameters_to_change):
```
Not sure, it is initializing the ArrayModel with all the required values / input. So it is doing what it is called for. Note that this is an internal function.

GernotMaier commented on simtools/model/array_model.py:
```diff
@@ -18,190 +16,189 @@ class InvalidArrayConfigData(Exception):
-        including reading the parameters from the DB.
+        Build the constituents of the array model (site, telescopes, etc).
+        Includes reading of all model parameters from the DB.
+        The array is define in the telescopes dictionary. Positions
```
Clarify that array element positions are read from the DB. Fixed the typo.

GernotMaier commented on simtools/model/array_model.py:
```diff
@@ -18,190 +16,189 @@ class InvalidArrayConfigData(Exception):
+            Site name.
+        parameters_to_change: dict
+            Dict with the parameters to be changed with respect to the DB model.
+
+        Returns
+        -------
+        SiteModel
+            Site model.
```
Fixed.

GernotMaier commented on simtools/model/array_model.py:
```diff
@@ -18,190 +16,189 @@ class InvalidArrayConfigData(Exception):
+                    telescope_model[element_name].change_multiple_parameters(**pars_to_change)
+                    telescope_model[element_name].set_extra_label(element_name)
+
+        return site_model, telescope_model
```
This PR is about telescope positions, calibration devices are handled elsewhere (see Tobias' work)

GernotMaier commented on simtools/schemas/model_parameters/array_element_position_ground.schema.yml:
```diff
@@ -0,0 +1,39 @@
+instrument:
+  class: Structure
+  type:
+    - LSTN
+    - LSTS
+    - MSTN
+    - MSTS
+    - SSTS
+    - SCTS
```
Yes, and this will be added with the first illuminator positions in the database (there are other array element positions to be added).

GernotMaier commented:
@VictorBarbosaMartins - thanks for the very thorough review.

I've think I've addressed all your comments. Some are beyond this PR and are (or has been already) moved to issues.

VictorBarbosaMartins commented on simtools/corsika/corsika_runner.py:
```diff
@@ -54,41 +52,30 @@ class CorsikaRunner:
-        Name of the layout.
+    array_model: ArrayModel
+        Array model instance.
+    simtel_source_path: str or Path
+        Location of source of the sim_telarray/CORSIKA package.
```
I meant by the [configurator module](https://github.com/gammasim/simtools/tree/main/simtools/configuration). As simtel_source_path might be defined as an env variable already.

GernotMaier replied:
I think this is used: the applications are using the configurator module and then pass this information along through the Simulator class. As we only need this part of the configuration, there is no need to pass the full configuration dict.

VictorBarbosaMartins commented:
> @VictorBarbosaMartins - thanks for the very thorough review.
> 
> I've think I've addressed all your comments. Some are beyond this PR and are (or has been already) moved to issues.

Thank you for addressing the comments. I am approving the PR as the rest of the issues will be addressed elsewhere.

VictorBarbosaMartins commented:
> Thank for this major update. I have mostly minor comments but quite a lot of them. I hope they help to improve the PR.
> 
> Point 1: In general, I have the impression that the ArrayLayout class, as we spoke, is becoming slowly obsolete (and with that the tool make_regular_arrays). I suggest opening an issue on that to focus on finalizing the transition to the array model class (without losing functionality).
> 
> Point 2: I suggest we already add the positions of the telescopes for prod5 to the DB as well. we have a stable environment at the moment running tests for prod5, but if we merge this branch without updating prod5 parameters, the tests will continue to be skipped. With this new db tool, it should be easy to do that, no?

Although I think I am missing a feedback on the two points I raised before.

GernotMaier commented:
1. The ArrayLayout class is not obsolete, as it provides the functionality to convert coordinates. Part of it might be replaced in future by a cta coordination transformation class, but this project has not even started. We will still need functionality to convert full layouts. There is a list of follow up issue coming from the changes to the handling of the telescope positions here: https://github.com/gammasim/simtools/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Telescope+positions+and+array+layout+using+the+database%22

2. Yes, the prod5 model needs to be completed at some point. But I also think we need to separate the work on the code and model, this are two different things.

GernotMaier commented:
Thanks again for the review and the useful comments. I will merge this branch then to main.
